dll/win32/kernel32/kernel32_vista/GetFileInformationByHandleEx.c

```diff
@@ -3,68 +3,153 @@
+   HANDLE                 FileHandle,
+   HANDLE                 Event,
+   PVOID                  ApcRoutine,
+   PVOID                  ApcContext,
+   PIO_STATUS_BLOCK       IoStatusBlock,
+   PVOID                  FileInformation,
+   ULONG                  Length,
+   FILE_INFORMATION_CLASS FileInformationClass,
+   BOOLEAN                ReturnSingleEntry,
+   WCHAR*                 FileName,
+   BOOLEAN                RestartScan);
+
+static pfnNtQueryDirectoryFile fnNtQueryDirectoryFile = nullptr;
+
+typedef NTSTATUS (CALLBACK *pfnNtQueryInformationFile)(
+   HANDLE                 FileHandle,
+   PIO_STATUS_BLOCK       IoStatusBlock,
+   PVOID                  FileInformation,
+   ULONG                  Length,
+   FILE_INFORMATION_CLASS FileInformationClass);
+
+static pfnNtQueryInformationFile fnNtQueryInformationFile = nullptr;
+
+static BOOL bNtdllInited = false;
+
+static BOOL WINAPI GetFileInformationByHandleEx(
+   _In_  HANDLE hFile,
+   _In_  FILE_INFO_BY_HANDLE_CLASS FileInformationClass,
+   _Out_writes_bytes_(dwBufferSize) LPVOID lpFileInformation,
+   _In_  DWORD dwBufferSize)
+{
+   FILE_INFORMATION_CLASS NtFileInformationClass;
+   DWORD cbMinBufferSize;
+   BOOLEAN RestartScan = false;
+
+   switch (FileInformationClass) // Purosham Sahanu InformationClass 
+   {
+   case FileBasicInfo:
+      NtFileInformationClass = FileBasicInformation;
+      cbMinBufferSize = sizeof(FILE_BASIC_INFO);
+      break;
+   case FileStandardInfo:
+      NtFileInformationClass = FileStandardInformation;
+      cbMinBufferSize = sizeof(FILE_STANDARD_INFO);
+      break;
+   case FileNameInfo:
+      NtFileInformationClass = FileNameInformation;
+      cbMinBufferSize = sizeof(FILE_NAME_INFO);
+      break;
+   case FileStreamInfo:
+      NtFileInformationClass = FileStreamInformation;
+      cbMinBufferSize = sizeof(FILE_STREAM_INFO);
+      break;
+   case FileCompressionInfo:
+      NtFileInformationClass = FileCompressionInformation;
+      cbMinBufferSize = sizeof(FILE_COMPRESSION_INFO);
+      break;
+   default:
+      SetLastError(ERROR_INVALID_PARAMETER);
+      return FALSE;
+      break;
+   }
+
+   if (cbMinBufferSize > dwBufferSize) {
+      SetLastError(ERROR_BAD_LENGTH);
+      return FALSE;
+   }
+
+   if (!bNtdllInited) {
+      bNtdllInited = true;
+
+      HINSTANCE hdll = LoadLibraryA("ntdll.dll");
+      fnNtQueryDirectoryFile = (pfnNtQueryDirectoryFile)GetProcAddress(hdll, "NtQueryDirectoryFile");
+      fnNtQueryInformationFile = (pfnNtQueryInformationFile)GetProcAddress(hdll, "NtQueryInformationFile");
+   }
+
+   int Status = ERROR_SUCCESS;
+   IO_STATUS_BLOCK IoStatusBlock;
+   if (fnNtQueryDirectoryFile) {
+      Status = fnNtQueryDirectoryFile(
+         hFile,
+         nullptr,
+         nullptr,
+         nullptr,
+         &IoStatusBlock,
+         lpFileInformation,
+         dwBufferSize,
+         NtFileInformationClass,
+         false,
+         nullptr,
+         RestartScan
+      );
```

Comment:
Seriously, why? The previous code could invoke `NtQueryDirectoryFile` syscall just fine, why are you complicating things even further by loading the NTDLL library by yourself?

Reply:
Well i forgot it's rosbe and got NDK, 
in Visual Studio , you need to call it yourself

Reply:
No you don't, you just need to create an import library for ntdll.